hal/src/common/gpio/v2/dynpin.rs

```diff
@@ -6,41 +6,40 @@
 //! value-level API solves this problem by erasing the type information and
 //! tracking the pin at run-time.
 //!
-//! Value-level pins are represented by the [DynPin] type. [DynPin] has two
-//! fields, `id` and `mode` with types [DynPinId] and [DynPinMode]
+//! Value-level pins are represented by the [`DynPin`] type. [`DynPin`] has two
```

Comment:
The term 'value-level' is unfamilar to me - is this the same as 'dynamically typed'?

Reply:
It's unfamiliar because I made it up. I thought it would help make things clear, but maybe I just made it worse? I was trying to contrast the `pin` module, which uses a "type-level" API, with the `dynpin` module, which uses a "value-level" API. I think maybe I explain the terms a bit more in the `v2` module docs.

At the value-level, a `DynPinMode` is a standard `enum`, just like you would normally write in Rust. Each different pin mode is represented as a different variant of the `DynPinMode` `enum`. But at the type-level, each different pin mode is its own type, and they are effectively variants of the `PinMode` meta-type (or type class), which is represented as a trait.

I couldn't find a way to summarize that distinction, so I made up the terms "type-level" and "value-level". But if it ends up adding rather than removing confusion, maybe I should find a different way to express it. What do you think?

I suppose you could call the "value-level" API a dynamically typed version of the "type-level" API. But that only makes sense if you expect pin modes to be represented as types to begin with. Outside of an embedded context, I don't think anyone would bother with the "type-level" approach. The "value-level" approach is much more idiomatic Rust.